compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParserErrorHandler.java

```diff
@@ -2730,6 +2730,8 @@ private ParserRuleContext getNextRuleForKeywords(ParserRuleContext currentCtx, i
                     return ParserRuleContext.CONFLICT_KEYWORD;
                 } else if (parentCtx == ParserRuleContext.QUERY_EXPRESSION) {
                     return ParserRuleContext.EXPRESSION;
+                } else if (parentCtx == ParserRuleContext.JOIN_CLAUSE) {
+                    return ParserRuleContext.EXPRESSION;
```

Comment:
This is related to above comment. Since we end the join-clause ctx for on-clause we don't need this if condition. 
next rule will be taken as expresssion under `QUERY_EXPRESSION` parent context in L2731-L2732